services/ExtraActivityManager.php

```diff
@@ -0,0 +1,352 @@
+            if ($value == 1) {
+                $previousTripples = $this->getTripplesForLearner($learnerName, $data['tag']);
+                if (!empty($previousTripples)) {
+                    $first = true;
+                    foreach ($previousTripples as $result) {
+                        if ($first) {
+                            $first = false;
+                            if (!in_array($this->tripleStore->update(
+                                $learnerName,
+                                self::LMS_TRIPLE_PROPERTY_NAME_EXTRA_ACTIVITY,
+                                $result['value'],
+                                json_encode($extraActivity),
+                                '',
+                                ''
+                            ), [0,3])) {// update
+                                if ($debug) {
+                                    $errorMessage .= 'Errors in '. get_class($this) . ' when updating '.$data['tag'].' for '.$learnerName .'<br>' ;
+                                }
+                                $error = true;
+                            }
+                        } else {
+                            // clean data
+                            $this->tripleStore->delete(
+                                $learnerName,
+                                self::LMS_TRIPLE_PROPERTY_NAME_EXTRA_ACTIVITY,
+                                $result['value'],
+                                '',
+                                ''
+                            );
+                        }
+                    }
+                } elseif ($this->tripleStore->create(
+                    $learnerName,
+                    self::LMS_TRIPLE_PROPERTY_NAME_EXTRA_ACTIVITY,
+                    json_encode($extraActivity),
+                    '',
+                    ''
+                ) > 0) {// create
+                    if ($debug) {
+                        $errorMessage .= 'Errors in '. get_class($this) . ' when creating '.$data['tag'].' for '.$learnerName .'<br>';
+                    }
+                    $error = true;
+                }
+                if (isset($oldExtraActivity)) {
+                    $oldExtraActivity->removeLearnerName($learnerName);
+                }
+            }
+        }
+        // remove old
+        if (isset($oldExtraActivity)) {
+            foreach ($oldExtraActivity->getRegisteredLearnerNames() as $learnerName) {
+                try {
+                    if (!$this->deleteExtraActivity($data['tag'], $learnerName)) {
+                        throw new \Exception('Errors in '. get_class($this) . ' when deleting '.$data['tag'].' for '.$learnerName .'<br>', 1);
+                    }
+                } catch (Throwable $t) {
+                    if ($debug) {
+                        $errorMessage .= $t->getMessage();
+                    }
+                    $error = true;
+                }
+            }
+        }
```

Comment:
j'ai mis du temps à comprendre comment se faisait la mise à jour. Ce qui relève de oldExtraActivity n'est pas très clair. Ca serait bcp plus simple de supprimer tous les tuples qui sont relatifs au tag puis de les rajouter avec un create en itérant les registeredLearnerNames. Si tu préfères garder comme ça, il faudrait au moins diviser en sous-fonctions et rajouter des commentaires.

Reply:
J'ai résolu la conversation car j'ai largement simplifié le code en question (ExtraACtivityManagerL131)